wagtail/core/models.py

```diff
@@ -1614,6 +1614,41 @@ def for_page(self, page):
         permission to perform specific tasks on the given page"""
         return PagePermissionTester(self, page)
 
+    def explorable_pages(self):
+        """Return a queryset of pages that the user has access to view in the
+        explorer (e.g. add/edit/publish permission). Includes all pages with
+        specific group permissions and also the ancestors of those pages (in
+        order to enable navigation in the explorer)"""
+        # Deal with the trivial cases first...
+        if not self.user.is_active:
+            return Page.objects.none()
+        if self.user.is_superuser:
+            return Page.objects.all()
+
+        explorable_pages = Page.objects.none()
+
+        # Creates a union queryset of all objects the user has access to add,
+        # edit and publish
+        for perm in self.permissions.filter(
+            Q(permission_type="add")
+            | Q(permission_type="edit")
+            | Q(permission_type="publish")
+        ):
+            explorable_pages |= Page.objects.descendant_of(
+                perm.page, inclusive=True
+            )
+
+        # For all pages with specific permissions, add their ancestors as
+        # explorable. This will allow deeply nested pages to be accessed in the
+        # explorer. For example, in the hierarchy A>B>C>D where the user has
+        # 'edit' access on D, they will be able to navigate to D without having
+        # explicit access to A, B or C.
+        page_permissions = Page.objects.filter(group_permissions__in=self.permissions)
+        for page in page_permissions:
+            explorable_pages |= page.get_ancestors()
```

Comment:
Quick q on this. You're correct here, that line does return the full list of ancestors.

But in practice, in the explorer view, it somehow ends up showing the right thing - only the highest common ancestor for pages they have permission for.

For example, if the tree structure is:

```
├── Root <no access>
│   ├── Blog Index 1 <add/edit/publish/lock access>
│   ├── Blog Index 2 <no access>
│   ├── Blog Index 3 <add/edit/publish/lock access>
```

`<Page: Root>` is included in that `explorable_pages` queryset, but in the explorer & the API that powers it, the user who only has access to "Blog Index 1" and "Blog Index 3" sees this:

![image](https://user-images.githubusercontent.com/1016999/56604082-2b43a580-65cf-11e9-828b-c5d324de979b.png)

I haven't figured out yet where/why that's happening, but Wagtail seems to be able determine the highest common ancestor for which a user has permissions, and the admin knows to call `?child_of` on that page to power the explorer, so it doesn't seem to matter if there are superfluous pages in the `explorable_pages` queryset.

@gasman are there any downsides with having superfluous pages in `explorable_pages`? I'm mostly concerned about having to do permissions checks for all the pages (or queryset diffing explorable_pages against "all pages a user has access to") that will add unnecessary overhead to rendering the explorer.

Reply:
@mike-hearn It works in this case because we call `get_explorable_root_page` when rendering the menu item, to determine the starting point for the explorer:
https://github.com/wagtail/wagtail/blob/master/wagtail/admin/wagtail_hooks.py#L39

I'd rather not rely on this, though, because it means that we end up with logic in `wagtail.core` that only exists to support one particular admin component and won't work as intended if it's used in other contexts. I don't think that adding this filter will require any major changes to this logic or add any further overhead: it should just involve running `first_common_ancestor` on the queryset of pages where an explicit permission record exists, and adding a filter on either `depth__gte=root_page.depth` or `path__startswith=root_page.path` to the final queryset.

Reply:
@gasman Good call - I believe I implemented your suggestion with this commit https://github.com/wagtail/wagtail/pull/5069/commits/7d1f550cba03454467f299aa7087b5257df49445 so let me know if I'm missing anything else in this PR.